sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayData.java

```diff
@@ -0,0 +1,376 @@
+import java.util.Set;
+
+/**
+ * Static display metadata associated with a {@link PTransform}. Display data is useful for
+ * pipeline runner UIs and diagnostic dashboards to display details about {@link PTransform}s
+ * that make up a pipeline.
+ *
+ * <p>{@link PTransform}s specify their display data by implementing the {@link HasDisplayData}
+ * interface.
+ */
+public class DisplayData {
+  private static final DisplayData EMPTY = new DisplayData(Maps.<Identifier, Item<?>>newHashMap());
+  private final ImmutableMap<Identifier, Item<?>> entries;
+
+  private DisplayData(Map<Identifier, Item<?>> entries) {
+    this.entries = ImmutableMap.copyOf(entries);
+  }
+
+  /**
+   * Default empty {@link DisplayData} instance.
+   */
+  public static DisplayData none() {
+    return EMPTY;
+  }
+
+  /**
+   * Collect the {@link DisplayData} for a {@link PTransform}. This will traverse all subcomponents
+   * specified via {@link Builder#include} in the given transform.
+   */
+  public static DisplayData from(PTransform<?, ?> transform) {
+    checkNotNull(transform);
+    return InternalBuilder.forRoot(transform).build();
+  }
+
+  public Collection<Item<?>> items() {
+    return entries.values();
+  }
+
+  public Map<Identifier, Item<?>> asMap() {
+    return entries;
+  }
+
+  /**
+   * Utility to build up display metadata from a @{link PTransform} and its included
+   * subcomponents.
+   */
+  public static interface Builder {
+    /**
+     * Include display metadata from the specified subcomponent. For example, a {@link ParDo}
+     * transform includes display metadata from the encapsulated {@link DoFn}.
+     *
+     * @return A builder instance to continue to build in a fluent-style.
+     */
+    Builder include(HasDisplayData subComponent);
+
+    /**
+     * Register the given display metadata.
+     */
+    ItemBuilder add(String key, String value);
+
+    /**
+     * Register the given display metadata.
+     */
+    ItemBuilder add(String key, long value);
+
+    /**
+     * Register the given display metadata.
+     */
+    ItemBuilder add(String key, double value);
+
+    /**
+     * Register the given display metadata.
+     */
+    ItemBuilder add(String key, Instant value);
+
+    /**
+     * Register the given display metadata.
+     */
+    ItemBuilder add(String key, Duration value);
+
+    /**
+     * Register the given display metadata.
+     */
+    ItemBuilder add(String key, Class<?> value);
+  }
+
+  /**
+   * Utility to append optional fields to display metadata, or register additional display metadata
+   * items.
+   */
+  public static interface ItemBuilder extends Builder {
+    /**
+     * Add a human-readable label to describe the metadata field. A label is optional; if
+     * unspecified, UIs should display the metadata key to identify the display item.
+     */
+    ItemBuilder withLabel(String label);
+
+    /**
+     * Add a link URL to the registered display metadata. A link URL is optional and
```

Comment:
Both of these methods should specify that the affect the mostly recently added metadata field.